### ISSUE TITLE
IndexOp for array values

### ISSUE THREAD

**Issue by joeltio** [joeltio/bento-box#32]
This issue keeps track of design decisions for adding a index operation to protobuf to handle retrieving values from arrays.

The main things to figure out are:
- Mutating specific value in multi-dimensional array
- Retrieving value in multi-dimensional array

**Comment by joeltio** [comment]
## Proposal: Partial indexing and lens-like mutations

### 1. Lens-like mutations

All mutations will be done by first retrieving the value to mutate, then changing the value. This will correspond to:
```
mutateOp(retrieveOp(attrRef), newVal)
```
The purpose of this is to delegate referencing values all to the `retrieveOp`. So we only have `AttributeRef` in one place.

This design is based off functional lenses.

### 2. `ValueRef` over `AttributeRef`

The usages of `AttributeRef` (graph `mutateOp` and `retrieveOp`, services get and set attribute) often refer to a value instead of an attribute. We want to modify a certain value or retrieve a certain value, not so much that we intentionally want to change an attribute.

I propose to add a `ValueRef` in place of existing `AttributeRef` usages. The `ValueRef` will look something like:
```
message ValueRef {
  // Which component to retrieve the value
  AttributeRef attr = 1;
  // A list of indexes
  repeated int64 indexes = 2;
}
```
For an attribute that is a primitive (non-array), the indexes will be empty.

**Comment by mrzzy** [comment]
> All mutations will be done by first retrieving the value to mutate, then changing the value.

What happens when only one element out of a large array is updated? Would it cause a full rewrite of the entire array?

> 2. ValueRef over AttributeRef

Looks good :+1:  , although we should probably disambiguate between Array indexes and index in ICS.

**Comment by joeltio** [comment]
> > All mutations will be done by first retrieving the value to mutate, then changing the value.
> 
> What happens when only one element out of a large array is updated? Would it cause a full rewrite of the entire array?

Nope. As `mutateOp` calls `retrieveOp` which uses `ValueRef`, we can pinpoint the exact value to update in the array to update.

> Looks good +1 , although we should probably disambiguate between Array indexes and index in ICS.

Yeah, but it seems like only the engine needs to disambiguate for now. The protobuf definitions have no idea of an ICS index.

**Comment by mrzzy** [comment]
> Yeah, but it seems like only the engine needs to disambiguate for now. The protobuf definitions have no idea of an ICS index.
How about `attr_indexes` or `array_indexes`?

**Comment by joeltio** [comment]
Ok, I think `attr_indexes` will be better:

```message ValueRef {
  // Which component to retrieve the value
  AttributeRef attr = 1;
  // A list of indexes
  repeated int64 attr_indexes = 2;
}
```

**Comment by mrzzy** [comment]
Seems like `ValueRef` is destined to replace `AttributeRef`  so maybe it makes sense to merge them together:
```proto
message ValueRef {
  // id of the entity to the retrieve from
  uint32 entity_id = 1;
  // Name of the component type to retrieve from entity
  string component = 2;
  // Name of the attribute to retrieve from component
  string attribute = 3;
   // A list of indexes
  repeated int64 attr_indexes = 4;
}
```
>  indexes remain unset for primitive values.

**Comment by joeltio** [comment]
To make things easier to change for SDK side, the `mutateOp` will still use `ValueRef`
